src/mp4.zig

```diff
@@ -0,0 +1,212 @@
+const std = @import("std");
+const Allocator = std.mem.Allocator;
+const Metadata = @import("metadata.zig").Metadata;
+
+// Some atoms can be "full atoms", meaning they have an additional 4 bytes
+// for a version and some flags.
+const FullAtomHeader = struct {
+    version: u8,
+    flags: u24,
+};
+
+fn readFullAtomHeader(reader: anytype) !FullAtomHeader {
+    return FullAtomHeader{
+        .version = try reader.readByte(),
+        .flags = try reader.readIntBig(u24),
+    };
+}
+
+/// Every atom in a MP4 file has this fixed-size header
+const AtomHeader = struct {
+    /// the atom size (including the header size of 8 bytes)
+    size: u32,
+    /// the name or type
+    name: [4]u8,
+};
+
+fn readAtomHeader(reader: anytype, seekable_stream: anytype) !AtomHeader {
+    const size = switch (try reader.readIntBig(u32)) {
+        0 => blk: {
+            // a size of 0 means the atom extends to end of file
+
+            const remaining = (try seekable_stream.getEndPos()) - (try seekable_stream.getPos());
+
+            break :blk @intCast(u32, remaining);
+        },
+        1 => {
+            // a size of 1 means the atom header has an extended size field
+            // TODO: implement this if relevant ?
+
+            std.debug.panic("unhandled 64 bit size", .{});
+        },
+        else => |n| n,
+    };
+
+    var name: [4]u8 = undefined;
+    _ = try reader.readAll(&name);
+
+    return AtomHeader{
+        .size = size,
+        .name = name,
+    };
+}
+
+/// Generic data atom
+///
+/// Its value should be interpreted based on the type indicator and locale indicator.
+/// For our purposes it's probably enough to treat them as UTF-8 strings.
```

Comment:
I think it might make more sense to validate that the data is UTF-8 and skip anything that's not explicitly UTF-8. AFAICT that would mean verifying that the first byte of the type indicator is `0` and that the remaining `u24` is `1`.

- [Type Indicator](https://developer.apple.com/library/archive/documentation/QuickTime/QTFF/Metadata/Metadata.html#//apple_ref/doc/uid/TP40000939-CH1-SW28)
- [Well-Known Types](https://developer.apple.com/library/archive/documentation/QuickTime/QTFF/Metadata/Metadata.html#//apple_ref/doc/uid/TP40000939-CH1-SW34)

Reply:
I implemented this.

As far as I can tell with my mp4 files the well-known type is always 1 for metadata.